### PR TITLE
VPA - Quick OOM for all containers

### LINKED ISSUE #2739: VPA - Quick OOMs for all containers
There was a discussion about quick OOMs:
https://github.com/kubernetes/autoscaler/pull/2479

IIUC the agreement was:
- add quick OOM handling for more than 1 containers.
- only containers with scaling policies should be consider for quick OOM.
- quick OOM doesn't apply to injected sidecars.

With [sidecar being currently detected](https://github.com/kubernetes/autoscaler/issues/2726) we are good to go with enhancing quick OOMs.

### PR DESCRIPTION
**Pull request by krzysied**
- handling quick OOM for more than 1 containers.
- only containers with scaling policies should be consider for quick OOM.
- quick OOM doesn't apply to injected sidecars.

/fixes #2739

### CODE REVIEW

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go [hunk 1]
```diff
@@ -336,37 +336,131 @@ func TestDontUpdatePodWithOOMAfterLongRun(t *testing.T) {
...
+			annotation: map[string]string{},
+			want:       true,
+		},
+		{
+			name:       "container listed in VpaOvservedContainers annotation",
+			annotation: map[string]string{annotations.VpaObservedContainersLabel: containerName},
+			want:       true,
+		},
+		{
+			// Continers not liseted in VpaOvservedContainers annotation
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: nit: 
s/Continers/Containers
s/liseted/listed/

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go [hunk 2]
```diff
@@ -336,37 +336,131 @@ func TestDontUpdatePodWithOOMAfterLongRun(t *testing.T) {
...
+	}{
 		{
-			LastTerminationState: apiv1.ContainerState{
-				Terminated: &apiv1.ContainerStateTerminated{
-					Reason:     "OOMKilled",
-					FinishedAt: metav1.NewTime(timestampNow.Add(-1 * 3 * time.Minute)),
-					StartedAt:  metav1.NewTime(timestampNow.Add(-1 * 5 * time.Minute)),
-				},
-			},
+			name:       "not VpaOvservedContainers annotation",
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: nit: s/not/no/

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go [hunk 3]
```diff
@@ -336,37 +336,131 @@ func TestDontUpdatePodWithOOMAfterLongRun(t *testing.T) {
...
+			// shouldn't trigger the quick OOM.
+			name:        "ContainerScalingModeOff",
+			scalingMode: vpa_types.ContainerScalingModeOff,
+			want:        false,
+		},
+	}
+	for _, tc := range tests {
+		t.Run(fmt.Sprintf("test case: %s", tc.name), func(t *testing.T) {
+			resourcePolicy := &vpa_types.PodResourcePolicy{
+				ContainerPolicies: []vpa_types.ContainerResourcePolicy{
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: You might want to add a test for default policies:
ContainerResourcePolicy = {
						ContainerName: "*",
						Mode:          &tc.scalingMode,
					},

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go [hunk 4]
```diff
@@ -336,37 +336,131 @@ func TestDontUpdatePodWithOOMAfterLongRun(t *testing.T) {
...
-	// Pod is within the recommended range.
-	recommendation := test.Recommendation().WithContainer(containerName).
-		WithTarget("5", "").
-		WithLowerBound("1", "").
-		WithUpperBound("6", "").Get()
-
-	calculator.AddPod(pod, recommendation, timestampNow)
-	result := calculator.GetSortedPods(NewDefaultPodEvictionAdmission())
-	assert.Exactly(t, []*apiv1.Pod{}, result, "Pod shouldn't be updated")
+func TestDontUpdatePodWithOOM_ContainerScalingMode(t *testing.T) {
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: TestDontUpdateModWithOOM is no longer a good name for this testcase

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go [hunk 5]
```diff
@@ -336,37 +336,131 @@ func TestDontUpdatePodWithOOMAfterLongRun(t *testing.T) {
...
-	calculator := NewUpdatePriorityCalculator(
-		nil, nil, &UpdateConfig{MinChangePriority: 0.5}, &test.FakeRecommendationProcessor{})
-
-	pod := test.Pod().WithName("POD1").AddContainer(test.BuildTestContainer(containerName, "4", "")).Get()
-
-	// Pretend that the test pod started 11 hours ago.
-	timestampNow := pod.Status.StartTime.Time.Add(time.Hour * 11)
-
-	pod.Status.ContainerStatuses = []apiv1.ContainerStatus{
+func TestDontUpdatePodWithOOM_VpaOvservedContainers(t *testing.T) {
```
Comment (bskiba) [inline vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: TestDontUpdateModWithOOM is no longer a good name for this testcase

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go [hunk 6]
```diff
@@ -85,23 +85,35 @@ func (calc *UpdatePriorityCalculator) AddPod(pod *apiv1.Pod, recommendation *vpa
...
+
 	updatePriority := calc.getUpdatePriority(pod, processedRecommendation)
 
 	quickOOM := false
-	if len(pod.Status.ContainerStatuses) == 1 {
-		terminationState := pod.Status.ContainerStatuses[0].LastTerminationState
+	for i := range pod.Status.ContainerStatuses {
+		cs := &pod.Status.ContainerStatuses[i]
+		if hasObservedContainers && !vpaContainerSet.Has(cs.Name) {
+			// Containers not observed by Admission Controller are not supported.
```
[the remark this answers is not in the document]
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go]: Done.

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go [hunk 7]
```diff
@@ -85,23 +85,35 @@ func (calc *UpdatePriorityCalculator) AddPod(pod *apiv1.Pod, recommendation *vpa
...
+			continue
+		}
+		crp := vpa_api_util.GetContainerResourcePolicy(cs.Name, calc.resourcesPolicy)
+		if crp != nil && crp.Mode != nil && *crp.Mode == vpa_types.ContainerScalingModeOff {
+			// Containers with ContainerScalingModeOff are not considered.
```
[the remark this answers is not in the document]
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator.go]: Done.

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go (hunk 1 above)
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: Done.

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go (hunk 2 above)
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: Done.

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go (hunk 3 above)
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: Tests added for `ContainerScalingModeAuto` and `ContainerScalingModeOff` as default.

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go (hunk 4 above)
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: Changed to `TestQuickOOM`.

File: vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go (hunk 5 above)
Comment (krzysied) [reply vertical-pod-autoscaler/pkg/updater/priority/update_priority_calculator_test.go]: Changed to `TestQuickOOM".

**Pull request merged**